### PR TITLE
 layers respect defaultVisibility if not defined in config

### LINKED ISSUE #4020: Load of REST dynamic service default
ESRI REST Dynamic service may be heavy and clients publish their ESRI REST service without displaying all the default layers. It would be a useful feature if when viewer load it reads the "REST" criterion "Default Visibility" to know if it should be true or false.

### PR DESCRIPTION
**Pull request by RyanCoulsonCA**
Closes #4020 

Layers that do not have a visibility defined in their configuration will now set their default visibility to the value retrieved from the server.

I tested this for both dynamic and feature layers by forcing defaultVisibility to false in the source code, but if anyone knows any layers that actually have defaultVisibility set to false for testing, please post below!

You can find a demo for this PR [here](https://fgpv-vpgf.github.io/fgpv-vpgf/fix-4020/samples/index-samples.html?sample=6).

- - -
This change is [Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039)

### CODE REVIEW

**Comment by james-rae** [comment]
From the issue analysis HackMD

- Remove config defaulting for layer properties, probably `visiblity` and `opaciy`. Allow `undefined` to be a config value post-parsing
- In layer load functions, need to recognize the `undefined` value. Set some internal flag or whatnot.
- After layer metadata loads, check the flag. If set, grab server setting result and apply to layer.
- Will need to handle children of Dynamics as well as just root level layers.
- Skip for File, WFS, WMS(?)

Was any of this considered?

**Comment by james-rae** [comment]
Because as is, appears to me the original problem persists: The config defaults all the booleans prior to load, so you never know if the user did or didn't provide a value

**Comment by james-rae** [comment]
Or is `config.source` like an untouched/undefaulted copy?

**Comment by james-rae** [comment]
*[`packages/ramp-geoapi/src/layer/layerRec/attribFC.js` line 72 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NC1FGTH75_-8e5bk96J:-NC1FGTH75_-8e5bk96K:b9sw06f) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/5f9e2d5dca9b2dc35a06487cad20f7af01ce10a2/packages/ramp-geoapi/src/layer/layerRec/attribFC.js#L72)):*
> ```JavaScript
>                   return data.defaultVisibility;
>               })
>             : true;
> ```

you prob want `Promise.resolve(true)`; right now this getter is returning Promise or boolean depending on state; might be some of your problems is because you're doing `true.then()` in reality.

**Comment by james-rae** [comment]
*[`packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js` line 389 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NC1Gkd22bkCVWNZv0wy:-NC1Gkd22bkCVWNZv0wz:b-vmfq37) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/5f9e2d5dca9b2dc35a06487cad20f7af01ce10a2/packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js#L389)):*
> ```JavaScript
>                     dFC.defaultVisibility.then((data) => {
>                         if (!hasVisibleState && data !== undefined) {
>                             // This isn't working, probably not finished loading
> ```

your prob right about the loading, i'mma look into the code guts.  But at a glance your line below could just be `dFC.setVisibility(data)`

**Comment by james-rae** [comment]
*[`packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js` line 428 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NC1I_Dt3G0qhiI-LwPj:-NC1I_Dt3G0qhiI-LwPk:b-gibpc1) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/5f9e2d5dca9b2dc35a06487cad20f7af01ce10a2/packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js#L428)):*
> ```JavaScript
>         });
> 
>         // TODO careful now, as the dynamicFC.DynamicFC constructor also appears to be setting visibility on the parent.
> ```

I think this block is where you need additional sauce.
We're using [this badboy](https://developers.arcgis.com/javascript/3/jsapi/arcgisdynamicmapservicelayer-amd.html#setvisiblelayers) to set the initial sublayer visibilities. Right now this code is synchronous; which was fine before since all our visible bools were known from the config. But now you have checkers above that are getting set AFTER the server spits back info about each child layer; so it happens after this code has already run.

Off the cuff idea.
This whole `if/else` gets put behind `const coulsonPromise = Promise.all(loadPromises).then(this code)`
Then you need to add `coulsonPromise` to the final `Promise.all([loadPromises, coulsonPromise])` at the bottom, to ensure the layer doesn't go to loaded until this `if/else` finishes.

Secondly, make a temporary data structure in this meatball of a function (`onLoad`) to track the boolean states of visibility. You can initialize that structure with some of the guts below (that just checks configuration as is); but then you will be updating this structure with your findings above, where you're calling `defaultVisibility()` and such.

Then it's your new structure that is used to figure out what gets passed to `_layer.setVisibility()` below.

**Comment by james-rae** [comment]
*[`packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js` line 428 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NC1I_Dt3G0qhiI-LwPj:-NC1K60Q26yoqNkWlh2m:b-m8c4ao) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/5f9e2d5dca9b2dc35a06487cad20f7af01ce10a2/packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js#L428)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

I think this block is where you need additional sauce.
We're using [this badboy](https://developers.arcgis.com/javascript/3/jsapi/arcgisdynamicmapservicelayer-amd.html#setvisiblelayers) to set the initial sublayer visibilities. Right now this code is synchronous; which was fine before since all our visible bools were known from the config. But now you have checkers above that are getting set AFTER the server spits back info about each child layer; so it happens after this code has already run.

Off the cuff idea.
This whole `if/else` gets put behind `const coulsonPromise = Promise.all(loadPromises).then(this code)`
Then you need to add `coulsonPromise` to the final `Promise.all([loadPromises, coulsonPromise])` at the bottom, to ensure the layer doesn't go to loaded until this `if/else` finishes.

Secondly, make a temporary data structure in this meatball of a function (`onLoad`) to track the boolean states of visibility. You can initialize that structure with some of the guts below (that just checks configuration as is); but then you will be updating this structure with your findings above, where you're calling `defaultVisibility()` and such.

Then it's your new structure that is used to figure out what gets passed to `_layer.setVisibility()` below.

</blockquote></details>

So in a nutshell, the flow is

1. Create an appropriate visibility tracking structure (index mapped to boolean?).
2. Init it BEFORE the `update async values` loop above. The init code is what we see below, checking `config.state.visibility` and such.
3. Then the async values code comes; almost as you have it now, but instead of setting visiblity on `dFC.setVisibility()`, you are updating your new structure.
4. The calls to `_layer.` in the `if/else` below then sit behind a promise.all that is waiting for all the other stuff (mainly the FC async calls, which are inspecting server visibility). 
5. Once that promise all resolves, we use your fancy structure to feed into the layer visibility initiation methods.  That code then resolves and our layer flips to loaded

**Review by RyanCoulsonCA — COMMENTED** [review]
> Or is config.source like an untouched/undefaulted copy?

Yeah, as far as I can tell `config.source` is basically an un-defaulted copy, at least for feature layers. Dynamic layer leaf nodes don't seem to have this.

> From the issue analysis HackMD

Totally forgot there was a HackMD, my bad. It seems like because config.source is an unmodified version of the properties my code sort of does the same thing anyway. I will look into removing the defaults though, since it doesn't seem like dynamic leaf nodes have this property.

> *[Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-:-NC1O4h-Bez7cwHqQDzg:b-k7u56w)* status: 0 of 6 files reviewed, 3 unresolved discussions (waiting on @james-rae, @RyanCoulsonCA, and @yileifeng)

___
*[`packages/ramp-geoapi/src/layer/layerRec/attribFC.js` line 72 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NC1FGTH75_-8e5bk96J:-NC1Q9XJ2y_myeWNu-51:b-vlm2yz) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/5f9e2d5dca9b2dc35a06487cad20f7af01ce10a2/packages/ramp-geoapi/src/layer/layerRec/attribFC.js#L72)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

you prob want `Promise.resolve(true)`; right now this getter is returning Promise or boolean depending on state; might be some of your problems is because you're doing `true.then()` in reality.
</blockquote></details>

Good catch :thumbsup: 
___
*[`packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js` line 428 at r1](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NC1I_Dt3G0qhiI-LwPj:-NC1Pglw9XimpKkdUJsf:b-fy0w78) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/5f9e2d5dca9b2dc35a06487cad20f7af01ce10a2/packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js#L428)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

So in a nutshell, the flow is

1. Create an appropriate visibility tracking structure (index mapped to boolean?).
2. Init it BEFORE the `update async values` loop above. The init code is what we see below, checking `config.state.visibility` and such.
3. Then the async values code comes; almost as you have it now, but instead of setting visiblity on `dFC.setVisibility()`, you are updating your new structure.
4. The calls to `_layer.` in the `if/else` below then sit behind a promise.all that is waiting for all the other stuff (mainly the FC async calls, which are inspecting server visibility). 
5. Once that promise all resolves, we use your fancy structure to feed into the layer visibility initiation methods.  That code then resolves and our layer flips to loaded

</blockquote></details>

This is super helpful, thanks. I'm going to play around and see if I can get it working.

**Comment by james-rae** [comment]
> since it doesn't seem like dynamic leaf nodes have this property

You'd probably need to use the parent layers `config.source.layerEntries[index].state.visibility` lol. I'm ok with the `.source` approach; this codebase is at a "patch only" phase of life so uber-elegance isn't required unless we need to. Your call.

**Comment by james-rae** [comment]
^ a bit wrong
`config.source.layerEntries.find(le => le.index === targetIndex).state.visibility`, with maybe extra code incase the `.find` returns nothing.

**Comment by RyanCoulsonCA** [comment]
I've made a bunch of changes to the PR and it works for both feature layers and dynamic layers now. Everything else seems to work the way it used to, but please look it over when you have time to make sure nothing has broken!

**Review by spencerwahl — APPROVED** [review]
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039)* status: all files reviewed, 3 unresolved discussions (waiting on @james-rae, @RyanCoulsonCA, and @yileifeng)

**Review by james-rae — CHANGES_REQUESTED** [review]
> Reviewed 4 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039)* status: all files reviewed, 5 unresolved discussions (waiting on @james-rae, @RyanCoulsonCA, and @yileifeng)

___
*[`packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js` line 472 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NCVUcvi9Bh0FqpxK-Iz:-NCVUcvi9Bh0FqpxK-J-:b23mqhj) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2390336dd87f8eeff39a6f9025d205ba177400a0/packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js#L472)):*
> ```JavaScript
>                     this.name = serviceResult.mapName;
>                 });
>                 loadPromises.push(setTitle);
> ```

mild concern with this. The Promise.all below is inspecting the `loadPromises` array. That `.all` call happens before this runs.  I don't know if you push a promise to the array after the fact, if the `.all` will consider it.
I'd say either remove this `if` from inside the `loadLayerPromise` guts (so the push just happens in the main function, and before the final Promise.all,  or you need to wire in the `setTitle` promise to act as a resolver to `loadLayerPromise`, which is probably more work.
___
*[`packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js` line 476 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NCVVcrb488QFgfb_bUr:-NCVVcrb488QFgfb_bUs:beg24al) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2390336dd87f8eeff39a6f9025d205ba177400a0/packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js#L476)):*
> ```JavaScript
>         });
> 
>         Promise.all([loadLayerPromise, loadPromises]).then(() => {
> ```

`loadPromises` is an array. so you're parameter is `[promise, [promise, promise, promise, ...]]`

You might want to do something like `loadPromises.concat([loadLayerPromise])`, or `[loadLayerPromise, ...loadPromises]`.

A `.push()` might work, but then your pushing `loadLayerPromise` into the array that `loadLayerPromise` is `.all`ing on which might create a black hole or something.
___
*[`packages/ramp-geoapi/src/layer/layerRec/layerRecord.js` line 300 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NCVXbY6-f8lCf30w44C:-NCVXbY6-f8lCf30w44D:b-x8nkad) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2390336dd87f8eeff39a6f9025d205ba177400a0/packages/ramp-geoapi/src/layer/layerRec/layerRecord.js#L300)):*
> ```JavaScript
> 
>         this.extent = shared.makeSafeExtent(this.extent);
>         this.defaultVisibility = this._layer.defaultVisibility;
> ```

this is becoming a layer property. I don't see it being read anywhere other than this function. It also has same name as the FC property but different data type (boolean vs Promise(boolean)).

If we don't need it outside of here, can this just be a `const`?
If we do need it persisted, can we add a comment about its type and whatnot. Without typescript this will get confusing.

**Review by RyanCoulsonCA — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039)* status: 4 of 6 files reviewed, 5 unresolved discussions (waiting on @james-rae, @RyanCoulsonCA, @spencerwahl, and @yileifeng)

___
*[`packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js` line 472 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NCVUcvi9Bh0FqpxK-Iz:-NCVbZU68seRgrK1q9ie:b-zcyad4) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2390336dd87f8eeff39a6f9025d205ba177400a0/packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js#L472)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

mild concern with this. The Promise.all below is inspecting the `loadPromises` array. That `.all` call happens before this runs.  I don't know if you push a promise to the array after the fact, if the `.all` will consider it.
I'd say either remove this `if` from inside the `loadLayerPromise` guts (so the push just happens in the main function, and before the final Promise.all,  or you need to wire in the `setTitle` promise to act as a resolver to `loadLayerPromise`, which is probably more work.
</blockquote></details>

You're right, didn't catch that. I've moved it outside of the loadLayerPromise.
___
*[`packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js` line 476 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NCVVcrb488QFgfb_bUr:-NCVbea30Zky_SRFXcOK:bt2vr0y) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2390336dd87f8eeff39a6f9025d205ba177400a0/packages/ramp-geoapi/src/layer/layerRec/dynamicRecord.js#L476)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

`loadPromises` is an array. so you're parameter is `[promise, [promise, promise, promise, ...]]`

You might want to do something like `loadPromises.concat([loadLayerPromise])`, or `[loadLayerPromise, ...loadPromises]`.

A `.push()` might work, but then your pushing `loadLayerPromise` into the array that `loadLayerPromise` is `.all`ing on which might create a black hole or something.
</blockquote></details>

Fixed!
___
*[`packages/ramp-geoapi/src/layer/layerRec/layerRecord.js` line 300 at r2](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039#-NCVXbY6-f8lCf30w44C:-NCVbfGI-i_S-XdbylpW:b-28bwbm) ([raw file](https://github.com/fgpv-vpgf/fgpv-vpgf/blob/2390336dd87f8eeff39a6f9025d205ba177400a0/packages/ramp-geoapi/src/layer/layerRec/layerRecord.js#L300)):*
<details><summary><i>Previously, james-rae (James Rae) wrote…</i></summary><blockquote>

this is becoming a layer property. I don't see it being read anywhere other than this function. It also has same name as the FC property but different data type (boolean vs Promise(boolean)).

If we don't need it outside of here, can this just be a `const`?
If we do need it persisted, can we add a comment about its type and whatnot. Without typescript this will get confusing.
</blockquote></details>

It definitely doesn't need to be a layer property. I've changed it to be a const.

**Review by yileifeng — APPROVED** [review]
> Reviewed 2 of 6 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @RyanCoulsonCA)

**Review by spencerwahl — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/fgpv-vpgf/fgpv-vpgf/4039)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @RyanCoulsonCA)
